File: src/main/java/cn/com/edtechhub/workcollaborativeimages/controller/UserController.java
package cn.com.edtechhub.workcollaborativeimages.controller;

import cn.com.edtechhub.workcollaborativeimages.annotation.LogParams;
import cn.com.edtechhub.workcollaborativeimages.exception.CodeBindMessageEnums;
import cn.com.edtechhub.workcollaborativeimages.enums.UserRoleEnums;
import cn.com.edtechhub.workcollaborativeimages.model.dto.UserTokenStatus;
import cn.com.edtechhub.workcollaborativeimages.model.entity.User;
import cn.com.edtechhub.workcollaborativeimages.model.request.userService.*;
import cn.com.edtechhub.workcollaborativeimages.model.vo.UserVO;
import cn.com.edtechhub.workcollaborativeimages.response.BaseResponse;
import cn.com.edtechhub.workcollaborativeimages.response.TheResult;
import cn.com.edtechhub.workcollaborativeimages.service.UserService;
import cn.com.edtechhub.workcollaborativeimages.utils.DeviceUtils;
import cn.com.edtechhub.workcollaborativeimages.utils.ThrowUtils;
import cn.dev33.satoken.annotation.SaCheckLogin;
import cn.dev33.satoken.annotation.SaCheckRole;
import cn.dev33.satoken.annotation.SaIgnore;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import io.swagger.v3.oas.annotations.Operation;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;

import javax.annotation.Resource;
import javax.servlet.http.HttpServletRequest;

/**
 * 用户控制层
 *
 * @author <a href="https://github.com/limou3434">limou3434</a>
 */
@RestController // 返回值默认为 json 类型
@RequestMapping("/user")
@Slf4j
public class UserController { // 通常控制层有服务层中的所有方法, 并且还有组合而成的方法, 如果组合的方法开始变得复杂就会封装到服务层内部

    /**
     * 注入用户服务实例
     */
    @Resource
    private UserService userService;

    /// 管理接口 ///

    @Operation(summary = "👑添加用户网络接口")
    @SaCheckLogin
    @SaCheckRole("admin")
    @PostMapping("/admin/add")
    @LogParams
    public BaseResponse<User> adminUserAdd(@RequestBody UserAddRequest userAddRequest) {
        User user = userService.userAdd(userAddRequest);
        return TheResult.success(CodeBindMessageEnums.SUCCESS, user);
    }

    @Operation(summary = "👑删除用户网络接口")
    @SaCheckLogin
    @SaCheckRole("admin")
    @PostMapping("/admin/delete")
    public BaseResponse<Boolean> adminUserDelete(@RequestBody UserDeleteRequest userDeleteRequest) {
        Boolean result = userService.userDelete(userDeleteRequest);
        return TheResult.success(CodeBindMessageEnums.SUCCESS, result);
    }

    @Operation(summary = "👑修改用户网络接口")
    @SaCheckLogin
    @SaCheckRole("admin")
    @PostMapping("/admin/update")
    public BaseResponse<User> adminUserUpdate(@RequestBody UserUpdateRequest userUpdateRequest) {
        return TheResult.success(CodeBindMessageEnums.SUCCESS, userService.userUpdate(userUpdateRequest));
    }

    @Operation(summary = "👑查询用户网络接口")
    @SaCheckLogin
    @SaCheckRole("admin")
    @PostMapping("/admin/search")
    public BaseResponse<Page<User>> adminUserSearch(@RequestBody UserSearchRequest userSearchRequest) {
        return TheResult.success(CodeBindMessageEnums.SUCCESS, userService.userSearch(userSearchRequest));
    }

    @Operation(summary = "👑封禁用户网络接口")
    @SaCheckLogin
    @SaCheckRole("admin")
    @PostMapping("/admin/disable")
    public BaseResponse<Boolean> adminUserDisable(@RequestBody UserDisableRequest userDisableRequest) {
        Boolean result = userService.userDisable(userDisableRequest.getId(), userDisableRequest.getDisableTime(), UserRoleEnums.USER_ROLE);
        return TheResult.success(CodeBindMessageEnums.SUCCESS, result);
    }

    @Operation(summary = "👑获取指定用户凭证网络接口")
    @SaCheckLogin
    @SaCheckRole("admin")
    @GetMapping("/admin/token")
    public BaseResponse<UserTokenStatus> adminGetUserToken(Long id) {
        UserTokenStatus userTokenStatus = userService.userGetTokenById(id);
        return TheResult.success(CodeBindMessageEnums.SUCCESS, userTokenStatus);
    }

    /// 普通接口 ///

    @Operation(summary = "用户注册网络接口")
    @SaIgnore
    @PostMapping("/register")
    public BaseResponse<Boolean> userRegister(@RequestBody UserRegisterRequest userRegisterRequest) {
        // 检查参数
        ThrowUtils.throwIf(userRegisterRequest == null, CodeBindMessageEnums.PARAMS_ERROR, "请求体不能为空");
        String account = userRegisterRequest.getAccount();
        String passwd = userRegisterRequest.getPasswd();
        String checkPasswd = userRegisterRequest.getCheckPasswd();
        ThrowUtils.throwIf(!passwd.equals(checkPasswd), CodeBindMessageEnums.PARAMS_ERROR, "两次输入的密码不一致");

        // 接口实现
        userService.userAdd(
                new UserAddRequest()
                        .setAccount(account)
                        .setPasswd(passwd)
        );
        return TheResult.success(CodeBindMessageEnums.SUCCESS, true);
    }

    @Operation(summary = "用户登入网络接口")
    @SaIgnore
    @PostMapping("/login")
    public BaseResponse<UserVO> userLogin(@RequestBody UserLoginRequest userLoginRequest, HttpServletRequest request) {
        User user = userService.userLogin(userLoginRequest.getAccount(), userLoginRequest.getPasswd(), DeviceUtils.getRequestDevice(request)); // 这里同时解析用户的设备, 以支持同端互斥
        UserVO userVO = UserVO.removeSensitiveData(user);
        return TheResult.success(CodeBindMessageEnums.SUCCESS, userVO);
    }

    @Operation(summary = "用户登出网络接口")
    @SaCheckLogin
    @PostMapping("/logout")
    public BaseResponse<Boolean> userLogout(HttpServletRequest request) {
        Boolean result = userService.userLogout(DeviceUtils.getRequestDevice(request));
        return TheResult.success(CodeBindMessageEnums.SUCCESS, result);
    }

    @Operation(summary = "获取登录凭证网络接口")
    @SaIgnore
    @GetMapping("/get/token")
    public BaseResponse<UserTokenStatus> userGetToken() {
        UserTokenStatus userTokenStatus = userService.userGetTokenById(userService.userGetCurrentLonginUserId());
        return TheResult.success(CodeBindMessageEnums.SUCCESS, userTokenStatus);
    }

    @Operation(summary = "获取登录会话网络接口")
    @SaIgnore
    @GetMapping("/get/session")
    public BaseResponse<UserVO> userGetSession() {
        User user = userService.userGetSessionById(this.userService.userGetCurrentLonginUserId());
        return TheResult.success(CodeBindMessageEnums.SUCCESS, UserVO.removeSensitiveData(user));
    }

}

File: client/work-collaborative-images-vue/src/api/work-collaborative-images/userController.ts
// @ts-ignore
/* eslint-disable */
import request from '@/request'

/** 👑添加用户网络接口 POST /user/admin/add */
export async function adminUserAdd(
  body: WorkCollaborativeImagesAPI.UserAddRequest,
  options?: { [key: string]: any }
) {
  return request<WorkCollaborativeImagesAPI.BaseResponseUser>('/user/admin/add', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 👑删除用户网络接口 POST /user/admin/delete */
export async function adminUserDelete(
  body: WorkCollaborativeImagesAPI.UserDeleteRequest,
  options?: { [key: string]: any }
) {
  return request<WorkCollaborativeImagesAPI.BaseResponseBoolean>('/user/admin/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 👑封禁用户网络接口 POST /user/admin/disable */
export async function adminUserDisable(
  body: WorkCollaborativeImagesAPI.UserDisableRequest,
  options?: { [key: string]: any }
) {
  return request<WorkCollaborativeImagesAPI.BaseResponseBoolean>('/user/admin/disable', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 👑查询用户网络接口 POST /user/admin/search */
export async function adminUserSearch(
  body: WorkCollaborativeImagesAPI.UserSearchRequest,
  options?: { [key: string]: any }
) {
  return request<WorkCollaborativeImagesAPI.BaseResponsePageUser>('/user/admin/search', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 👑获取指定用户凭证网络接口 GET /user/admin/token */
export async function adminGetUserToken(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: WorkCollaborativeImagesAPI.adminGetUserTokenParams,
  options?: { [key: string]: any }
) {
  return request<WorkCollaborativeImagesAPI.BaseResponseUserTokenStatus>('/user/admin/token', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** 👑修改用户网络接口 POST /user/admin/update */
export async function adminUserUpdate(
  body: WorkCollaborativeImagesAPI.UserUpdateRequest,
  options?: { [key: string]: any }
) {
  return request<WorkCollaborativeImagesAPI.BaseResponseUser>('/user/admin/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 获取登录会话网络接口 GET /user/get/session */
export async function userGetSession(options?: { [key: string]: any }) {
  return request<WorkCollaborativeImagesAPI.BaseResponseUserVO>('/user/get/session', {
    method: 'GET',
    ...(options || {}),
  })
}

/** 获取登录凭证网络接口 GET /user/get/token */
export async function userGetToken(options?: { [key: string]: any }) {
  return request<WorkCollaborativeImagesAPI.BaseResponseUserTokenStatus>('/user/get/token', {
    method: 'GET',
    ...(options || {}),
  })
}

/** 用户登入网络接口 POST /user/login */
export async function userLogin(
  body: WorkCollaborativeImagesAPI.UserLoginRequest,
  options?: { [key: string]: any }
) {
  return request<WorkCollaborativeImagesAPI.BaseResponseUserVO>('/user/login', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 用户登出网络接口 POST /user/logout */
export async function userLogout(options?: { [key: string]: any }) {
  return request<WorkCollaborativeImagesAPI.BaseResponseBoolean>('/user/logout', {
    method: 'POST',
    ...(options || {}),
  })
}

/** 用户注册网络接口 POST /user/register */
export async function userRegister(
  body: WorkCollaborativeImagesAPI.UserRegisterRequest,
  options?: { [key: string]: any }
) {
  return request<WorkCollaborativeImagesAPI.BaseResponseBoolean>('/user/register', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}
